test/mpi/shmem_pingpong.c

```diff
@@ -9,79 +9,185 @@
 #include <stdlib.h>
 #include <sys/time.h>
 #include <string.h>
+#include <getopt.h>
+#include <limits.h>
 #include <shmem.h>
 
 
-static inline void wait_until(void *shared_buffer, long sn)
+#define GLOBAL_DATA_SIZE   1024
+
+
+static int show_result(const struct timeval *tv_prev,
+                       const struct timeval *tv_curr,
+                       long iters, size_t msg_size, int force)
 {
-    if (0) {
-        while (*(volatile long*)shared_buffer != sn);
-    } else {
-        shmem_wait_until(shared_buffer, SHMEM_CMP_EQ, sn);
+    double elapsed;
+
+    elapsed = (tv_curr->tv_sec + tv_curr->tv_usec * 1e-6) -
+              (tv_prev->tv_sec + tv_prev->tv_usec * 1e-6);
+
+    if (((elapsed >= 1.0) || force) && (shmem_my_pe() == 0)) {
+        printf("%ld iterations, %lu bytes, latency: %.3f usec\n",
+               iters, msg_size, elapsed * 1e6 / iters / 2.0);
+        return 1;
     }
+
+    return 0;
 }
 
-static void run(void *shared_buffer)
+static void run_pingpong(char *mem, size_t msg_size, long num_iters, int use_wait,
+                         int do_quiet)
 {
-    const long iters = 100000;
-    struct timeval tv_start, tv_end;
-    double latency_usec;
-    long sn;
-    long i;
-
-    gettimeofday(&tv_start, NULL);
-
-    sn = 1;
-    if (shmem_my_pe() == 0) {
-        for (i = 0; i < iters; ++i) {
-            shmem_long_p(shared_buffer, sn, 1);
-            wait_until(shared_buffer, sn);
-            ++sn;
+    struct timeval tv_prev, tv_curr;
+    int my_pe, dst_pe;
+    volatile int *rsn;
+    char *msg;
+    int *ssn;
+    int sn;
+    long i, prev_i;
+
+    msg = malloc(msg_size);
+    if (msg == NULL) {
+        return;
+    }
+
+    memset(msg, 0, msg_size);
+
+    gettimeofday(&tv_prev, NULL);
+    prev_i = 0;
+    my_pe  = shmem_my_pe();
+    dst_pe = 1 - my_pe;
+    rsn    = (int*)&mem[msg_size - sizeof(int)];
+    ssn    = (int*)&msg[msg_size - sizeof(int)];
+
+    for (i = 0; i < num_iters; ++i) {
+        sn   = i & 127;
+        *ssn = sn;
+        if (my_pe == 0) {
+            shmem_putmem(mem, msg, msg_size, dst_pe);
+            if (do_quiet) {
+                shmem_quiet();
+            }
         }
-    } else {
-        for (i = 0; i < iters; ++i) {
-            wait_until(shared_buffer, sn);
-            shmem_long_p(shared_buffer, sn, 0);
-            ++sn;
+        if (use_wait) {
+            shmem_int_wait_until(rsn, SHMEM_CMP_EQ, sn);
```

Comment:
1. the shmem way of doing this: shmem_put(data); shmem_fence(); shmem_put(flag);
2. If the message size is equal to the datatype used in wait, you don't need the flag. Otherwise you want to do #1.

Having said this, for investigation of the latency on Mellanox IB devices the approach that you used is most useful, since it exposes the low level latency. I know for sure it does not work this way for other interconnects.

So it really depends what is the main goal of this utility...